src/translators/crate.py

```diff
@@ -475,13 +426,129 @@ def query(self,
+         'time_index': [t0, t1, ..., tn],
+         'attr_1': {
+             'index': [t0, t1, ..., tn], # index of this attr (if different)
+             'values': [v0, v1, ..., vn],
+             'type': Number
+         },
+         ...,
+         'attr_N': ...
+        },...
+        ]
+
+        It returns an array of dictionaries, each representing a query result
+        on a particular NGSI Entity Type. Each of the dicts in this array
+        consists of the following attributes.
+
+        'type' is the NGSI Entity Type of the response.
+
+        'id' or 'ids'. id if the response contains data from a specific NGSI
+        entity (with that id) or ids in the case the response aggregates data
+        from multiple entities (those with those ids). You get one or the
+        other, not both.
+
+        'time_index': The time index applying to the response, applies to all
+        attributes included in the response. It may not be present if each
+        attribute has its own time index array, in the cases where attributes
+        are measured at different moments in time. Note since this is a "global"
+        time index for the entity, it may contain some NULL values where
+        measurements were not available. It's an array containing time
+        measured in milliseconds since epoch (typically in the timezone the
+        Orion Notification used, or UTC if created within QL).
+
+        Each attribute in the response will be represented by a dictionary,
+        with an array called 'values' containing the actual historical values
+        of the attributes as queried. An attribute 'type' will have the original
+        NGSI type of the attribute (i.e, the type of each of the elements now
+        in the values array). The type of an attribute is not
+        expected to change in time, it'd be an error. Additionally, it may
+        contain an array called 'index', just like time_index but for this
+        specific attribute. Thus, this 'index' will never contain NONE values.
+
+        If the user did not specify an aggrMethod, the response will not mix
+        measurements of different entities in the same values array. So in this
+        case, there will be many dictionaries in the response array, one for
+        each NGSI Entity.
+
+        When using aggrPeriod, the index array is a completely new index,
+        composed of time steps of the original index of the attribute but
+        zeroing the less significant bits of time. For example, if there were
+        measurements in time 2018-04-03T08:15:15 and 2018-04-03T09:01:15, with
+        aggrPeriod = minute the new index will contain, at least, the steps
+        2018-04-03T08:15:00 and 2018-04-03T09:01:00 respectively.
+
+        :raises:
+        ValueError in case of misuse of the attributes.
+        UnsupportedOption for still-to-be-implemented features.
+        crate.DatabaseError in case of errors with CrateDB interaction.
```

Comment:
does this include time out? how do we handle very large queries? can we "segment"?

Reply:
In spite of the name, `DatabaseError` is wrapping timeouts.
Queries are capped by QL and CrateDB at [10000](https://crate.io/docs/crate/reference/en/latest/general/dql/selects.html#limits).
As suggested by Crate folks, the user could make use of `offset` and `limit` to deal with paging at the moment. QL is exposing these two params, and there are simple unit-tests for both, but ideally this would need further testing.